### ISSUE TITLE
Allow extensions to publish beta releases and users to opt-in to them

### ISSUE THREAD

**Issue by DanTup** [microsoft/vscode#15756]
Apologies if this has come up before; I'm unable to find any related issues.

It would be cool to be able to publish extensions in a beta form (similar to Code Insiders) so that they can be run "in production" with real users that have opted-in. This would reduce the disk in publishing an extension because you can push it to a smaller number of users for testing (I probably don't need to sell this to you, I guess you have Insiders for very similar reasons).

I know we can package and distribute extensions ourselves for now, but it's not as easy (eg. updating is not automatic).

I originally thought the "Preview" flag might've provided this functionality but unless I've overlooked something, it does not seem to.

**Comment by ramya-rao-a** [comment]
I would very much like this myself.

**Comment by joaomoreno** [comment]
Why not putting those experimental features behind a configuration setting and have the opt-in mechanism be toggling that setting to `true`? Once the feature is stable enough just remove the setting.

**Comment by DanTup** [comment]
> Why not putting those experimental features behind a configuration setting and have the opt-in mechanism be toggling that setting to `true`?

This is an option, but I think it has some drawbacks

1. If you want to do a beta release every day, your "stable" users will get the update notifications too
2. It doesn't eliminate all the risk (if you're shipping bugs in the new features, maybe your feature detection is buggy, or how you're dealing with two code paths)
3. For things that aren't just new features, it's more work to maintain/switch between the branches
4. Users have to keep opting-in to each feature, rather than choosing to be a beta tester once; this could really limit the testing you get
5. Having the feature toggles show up in the config auto-complete isn't ideal (this might not be an issue, but I didn't get a response to https://github.com/Microsoft/vscode/issues/1396#issuecomment-244350985 so it's unclear if "hidden options" are supposed to be used)

None of these issues are huge, but since the Insiders builds work so well and it seems like a similar scenario it seemed like a good request to raise for discussion :)

**Comment by davidanthoff** [comment]
I agree, this would be a really useful feature. Right now I'm distributing our beta releases via VSIX downloads, but that is cumbersome and I don't reach many folks that way.

Hiding things behind a feature flag wouldn't really help. For example, we are about to release a new version that changes a lot of code, the whole architecture of the extension. There is no way that we could have had that behind a feature flag.

**Comment by eamodio** [comment]
I would love to see this as well -- while feature flags work in certain cases, it doesn't cover them all. It also provides a sanity check and limits the "damage" of a bad release. And installing via vsix is just too user unfriendly to really be viable imo -- as it is completely manual.

**Comment by eamodio** [comment]
Just to kick this one again, because the issues surrounding the manual install and REQUIRED uninstall of vsix causes lots of pain for users and seemingly creates cases of corrupt installations -- I've definitely seen an up tick in GitLens issues caused by vsix/marketplace versions on top of one another.

**Comment by DanTup** [comment]
Yeah, I meant to post back here - when I did a beta release a few months ago I had reports of issues installing/uninstalling the vsix:

- https://github.com/Dart-Code/Dart-Code/issues/318#issuecomment-308244455
- https://github.com/Dart-Code/Dart-Code/issues/318#issuecomment-308263912
- https://github.com/Dart-Code/Dart-Code/issues/318#issuecomment-308279628

**Comment by DanTup** [comment]
I gave up on vsix testing because of the complications mentioned above. Yesterday I shipped a huge bug that only occurred if `dart` on your `PATH` was a symlink (I had to make a lot of changes for multi-root support, and putting it all behind flags was not feasible) :(

I'm starting to think about publishing a second version of my extension with Beta in the name to allow people to choose to test new releases earlier (seems like [at least one other](https://marketplace.visualstudio.com/items?itemName=webfreak.code-d-beta) has done this).

**Comment by sandy081** [comment]
@DanTup @eamodio We did some improvements in this milestone (November) to support beta releases in VS Code when installed through VSIX. Refer to this issue https://github.com/Microsoft/vscode/issues/39024 for more details. Can you please check if that helps you in using VSIX?

**Comment by DanTup** [comment]
@sandy081 Sounds good! Am I right in thinking november milestone is 1.19 and not released to stable yet? I'll try and do some testing latest this/next week.

**Comment by sandy081** [comment]
Yes 1.19 will be released to stable this week. Let me know if you still want this feature to be opened after that.

**Comment by DanTup** [comment]
@sandy081 Haven't had chance to test it yet; but I think there's still value in this feature even if vsix works better than it did, for example:

- I think end users would have more confidence installing a beta vsix through the marketplace then from some random download url
- Allowing people to opt-in to a beta channel would allow them to get auto-updates of betas instead of having to do manual installations all the time
- Listing betas on the marketplace is much more discoverable - only a fraction of my users will follow the twitter account (etc.), so there are a lot of potential testers that simply won't know about beta vsix releases

**Comment by sandy081** [comment]
@DanTup All your requirements are valid. But I think this has to be supported by the Market place first. 

@modigrv May I know if there are any plans to have this support in Market place?

**Comment by DanTup** [comment]
😔

**Comment by ramya-rao-a** [comment]
Re-opening this issue as we haven't reached any conclusion yet.
ping @seanmcbreen and @modigrv on the question on marketplace

**Comment by sean-mcmanus** [comment]
Our C/C++ extension wants to be able to publish insider builds to the marketplace so users who opt-in can automatically get either the "insider" version or the release version after that's published -- mostly to increase our real-world test coverage before pushing changes to everyone. We can publish the insider vsix to our GitHub site and email users about it who sign up for it, but it requires user to manually install it, so that isn't likely to get enough usage. We're fine with all VS Code Insiders users also getting the Insider extension instead of having a separate insider setting just for our extension.

**Comment by sandy081** [comment]
I do not think we can do anything from VS Code side without such support from Market place.

**Comment by davidanthoff** [comment]
Where can folks give feedback that this is important? I thought leaving this issue open so that people can still like it might provide that option?

**Comment by DanTup** [comment]
> We're fine with all VS Code Insiders users also getting the Insider extension instead of having a separate insider setting just for our extension.

I don't think it'd be great to tie these together since it'd mean using the insiders version of your extension means that the user needs to be happy to switch also to Insiders Code and with it any other extensions that are publishing Insiders versions. I wouldn't want to go "full unstable" just to do it for one extension (for one, the stability of Insiders extensions could vary wildly across projects).

It also changes the environment somewhat, so you're not getting an accurate test of what happens if you publish the extension (unless ofc you wait for the next stable Code release).

> I do not think we can do anything from VS Code side without such support from Market place.

Where can we show support for this? If this case stays closed, I suspect this will be forgotten about but it seems to be a good request desired by several extensions.

**Comment by sean-mcmanus** [comment]
Yeah, having extension "insider-ness" be independent of VS Code itself is fine with me too. We just need the marketplace to allow us to submit and insider version and then VS Code would need to add an "Install Insider" button next to extensions that also have an insider version. Does the marketplace have some feedback site where this is being tracked?

**Comment by eamodio** [comment]
They could even just re-use the `preview` flag -- I would bet users don't know what that is even supposed to mean (I don't -- other than some signal with regard to quality/stability/features -- but very nebulous and I bet doesn't stop bad reviews).

**Comment by DanTup** [comment]
I guess the Marketplace would still need explicit support for this since people who don't want the preview/insiders/pre-release versions would expect to browser the marketplace and see the release notes (etc.) for the latest stable release, not the preview release.

**Comment by eamodio** [comment]
Yeah, agreed -- there would definitely be a bunch of plumbing that they'd need to do on the marketplace side.

**Comment by sandy081** [comment]
@viradhamMS @pkuma-msft We have a popular request to support beta/insiders versions of extensions which in turn needs Market place support.

May I know if there are any plans to have this support in Market place and what is the right Market place forum to raise these requests?

Reopening on users demand :)

**Comment by viradhamMS** [comment]
Thanks folks. We do have a backlog feature to enable similar requirement. 
When we get to it, we will reach out to you folks to understand the requirements to finer detail.

For such feedback/questions on Marketplace, Please send a mail to vsmarketplace @ microsoft.com

**Comment by sean-mcmanus** [comment]
It seems like extensions themselves can implement their own insiders channel without marketplace support for users who have some "yourExtension.insiders" setting set:
1. Downloading some json file that has the current insiders version of the extension on activation (we already do this for on the fly a/b testing).
2. If there's a new insider version, the extension can automatically download/install the vsix.
And if you want more extension insider usage, you could set the default setting to true for VS Code insiders builds, and users could change the setting to false if they want the release version. This would hopefully solve our current problem in that our manually installed vsix has way too small 0.03% installs compared to the released version. It's on our backlog: https://github.com/Microsoft/vscode-cpptools/issues/1526 .

Theoretically, this mechanism could be implemented in a way that sharable between extensions via something like:
1. Adding a setting extensions.insiders: [ "ext1": true, "ext2": false].
2. Adding a property to package.json that sets the URL to download that points to a json file with the current insiders version and URL. We actually have 4 additional "offline" installer URLs for different platforms that wouldn't fit into this design, but we could ship just the online installer via this mechanism.

But I don't know how many other extensions would want to take advantage of this.

**Comment by eamodio** [comment]
@sean-mcmanus that sounds very interesting. Personally I would like for vscode to somewhat "approve" a mechanism like this before I'd employ it in my extensions.

**Comment by stkb** [comment]
While I was wanting this feature too, I have to say, the improved support for installing from .vsix files (without fear of conflicts) in #39024 is working nicely for me. Most people who have an issue with my extension are only interested in that one thing anyway. So I can attempt a fix, upload a beta .vsix file though github releases, and point them there. If everything's ok, I tell them to stick with that and then with the next marketplace release they're updated automatically and back on stable.

Now of course... it work for all cases; large extensions that want a permanent beta channel. Just wanted to share for anyone that might have missed that change.

**Comment by DanTup** [comment]
The vsix fix is good; but I've had several beta releases in the last week or so. Telling users to keep updating vsix's kinda sucks. Having auto-updates of betas would make is so much easier - I wouldn't have to convince people to go download another vsix and install it :)

**Comment by tylerl0706** [comment]
Hey everyone! I work on the PowerShell extension and would LOVE to have a preview/beta channel. In the meantime, I'm probably going to go down the route of having a separate extension [similar to what tslint does](https://marketplace.visualstudio.com/items?itemName=eg2.ts-tslint).

Shoutout to @mattbierner for pointing me here 👍

**Comment by sean-mcmanus** [comment]
FYI, the 0.19.0 C/C++ extension has implemented a C_Cpp.updateChannel setting with an "Insiders" option that will automatically check our GitHub releases page for a newer Insiders .vsix and download/install that: see https://github.com/Microsoft/vscode-cpptools/tree/release/Extension/src . Users of the Insiders VS Code have that setting defaulted to "Insiders".

**Comment by eamodio** [comment]
@sean-mcmanus That's very cool -- though I would still really like this to be built-in/provided functionality.

**Comment by davidanthoff** [comment]
Couldn't the ``preview`` tag in a published extension manifest be used for this? If a release has the ``preview`` property as ``true``, don't install it, unless a user opts into the "preview" channel for that extension. It seems to me that this wouldn't require any changes on the marketplace but could be implemented in VS Code itself?

**Comment by TylerLeonhardt** [comment]
Ideally we'd want to be able to grab the latest non-preview if we want the stable channel, or the latest preview if we want the preview channel.

**Comment by sean-mcmanus** [comment]
@davidanthoff "Preview" is currently being used by some extensions for their "release" versions that are not considered "feature complete" enough for 1.0, so you'd be changing the meaning of the "preview" field. It would be better to check the version to see if it contains an "-insiders<num>" which is what our extension uses to implement our `"C_Cpp.updateChannel": "Insiders"`.

**Comment by davidanthoff** [comment]
I see. In any case, I don’t understand why this can’t just be implemented in VS Code itself, why is there a need for a marketplace update?

**Comment by sean-mcmanus** [comment]
VS Code's extension API doesn't have a generic concept of "<extension-setting-prefix>.updateChannel", so they would have to add some package.json setting for extensions to provide what setting should be queried for that (or some other equivalent mechanism). If that is done, then they could use an implementation similar to what our C/C++ extension does to bypass the marketplace and query GitHub for new Insider vsix's.

**Comment by sean-mcmanus** [comment]
@davidanthoff How would VS Code determine is an extension is on a prerelease channel? There needs to be some API or package.json for setting that (e.g. an "updateChannel" or "updateChannelSettingName" field in package.json). The marketplace only allows submission/publishing of one version at a time -- we can't submit a 0.20.1 and 0.21.0-insiders without changes to the marketplace implementation (only the last submitted would be used).

**Comment by davidanthoff** [comment]
But aren't all versions that have ever been published stored on the marketplace? I just tried this with our julia package, and I can get any old version with a link like
````
https://marketplace.visualstudio.com/_apis/public/gallery/publishers/julialang/vsextensions/language-julia/0.8.0/vspackage
````
I guess the problem is that the extension can't get a list of all the uploaded extension versions?

**Comment by sean-mcmanus** [comment]
@davidanthoff Yeah, I think you're right about all the older versions being saved (we used to tell users to use that when there was some breaking regression), but VS Code would need a way to query the marketplace for the list (or the latest Insiders versus non-Insiders). It's possible such an web API exists, but I haven't checked. Our extension queries https://api.github.com/repos/Microsoft/vscode-cpptools/releases for the version list.

**Comment by TylerLeonhardt** [comment]
That API you linked has whether or not a version is a prerelease or not. Could we go down the list for the last non-prerelease?

Also couldn't help but notice cpptools is following semver so that could also be leveraged, potentially.

**Comment by sean-mcmanus** [comment]
@TylerLeonhardt Do you mean the GitHub API? That only works for our extension because we post our Insider .vsix's on GitHub and we have custom code that queries it and updates our extension with that. VS Code itself could do something similar to give other extensions the ability to have an Insider channel -- as mentioned previously, there just needs to be some mechanism for exposing the per-extension setting to turn the Insider mode on or off for an extension. FYI, our extension has gotten thousands of users on our Insider usage, so we consider it a success (compared to tens of users previously).

**Comment by TylerLeonhardt** [comment]
Oof my bad. Got confused on who is who in this thread and totally thought the API you linked was from the marketplace 😄

if the marketplace has an API similar to GitHub's... that could be leveraged in vscode to achieve the NuGet Package Manager-like experience of installing extensions - checkbox to allow prerelease/insiders/previews... which filters the list.

**Comment by sean-mcmanus** [comment]
@davidanthoff I'm not aware of a per-extension setting in VS Code -- maybe the "setting" could just be exposed via the right-click menu in the Extensions pane, like the "Install Another Version..." option is. Also, in order for us to switch to this VS Code implementation, the per-extension setting would need to default to "Insiders" for users of VSCode Insiders, because that's how we get most of our current Insider usage (and enable people who don't want to be on Insiders to change the default).

**Comment by TylerLeonhardt** [comment]
Implemented a workaround for this by shipping two extensions...

Preview
https://marketplace.visualstudio.com/items?itemName=ms-vscode.PowerShell-Preview

Stable
https://marketplace.visualstudio.com/items?itemName=ms-vscode.PowerShell

I ended up running into a strange issue trying to release a `2.0.0-preview.1` version of the Preview package:
![can't use 2.0.0-preview](https://user-images.githubusercontent.com/2644648/51644469-1cdce380-1f25-11e9-911c-9530357ac8ce.png)

No support for SemVer 2! :(

So I tried to use `2.0.0.0` instead and npm doesn't support 4 numbers... only SemVer 2 format!

It's been a wild ride... I really hope the code team can do something about this.

Right now extensions can only be X.X.X versioned.

**Comment by DanTup** [comment]
In some ways I like that I can't accidentally publish a dev release. I'm really nervous when I type `vsce package` that I might push a dev build to all users!

I wonder if we could still have a prefix that was vblocked (eg. `-dev`) even it became possible to upload other pre-release versions (in order to support this). There would need to be some work in VS Code and on the site to let users see/pick between stable/preview releases, but being able to see versions in VS Code now seems like a step towards this.

**Comment by beriberikix** [comment]
Out of curiosity, where does one file issues against the Marketplace? It sounds like we need to raise awareness to the folks who maintain the Marketplace as well as the vscode team (who are obviously already on this thread.)

**Comment by viradhamMS** [comment]
@beriberikix - You can file issues here itself, we work very closely with VSCode group and it will reach us pretty quickly (or) feel free to directly post to vsmarketplace at microsoft dot com

**Comment by natanfudge** [comment]
A feature like this would help out immensely for pushing new features. I'm hoping the marketplace team will get on it soon!

**Comment by sandy081** [comment]
Currently most of the extensions are following a separate extension for extension insiders/nightly. This way, users can opt in to install either stable/insiders version of extension and disable the other. It is also **NOT** limited to insiders/stable.

Any feedback on this solution?

**Comment by DanTup** [comment]
I don't think it's the best solution - it means you're listed twice in the marketplace, have two sets of reviews/stats, and have to fiddle around when publishing to change the extension ID.

It also means extensions that reference each other by ID (for example Dart & Flutter are dependent and call exported APIs) need some faffing around too (to know to check which version is installed).

And of course, we'd need to do our own checks that the user didn't install multiple versions.

I think being able to publish pre-release versions to the marketplace (even if it's a white-listed set, like `-beta`) and let users have a flag per-extension to opt in the using pre-release versions would be far better (and I don't think it should be tied to VS Code Insiders).

**Comment by sandy081** [comment]
Your concerns are valid.

Another approaches we are discussing which is based on what is suggested here

## Approach 2: Defining Insiders Flag (VS Code hosted solution)

- Introduce an insiders flag for extensions which can be provided in the `package.json` and follow the same versioning system `major.minor.patch`. 

- Provides a way (some UI) to the user to install the stable or opt in for insiders in VS Code. User can always switch between stable and insiders flavour of extension even after installing.

**Drawbacks**

- Following the same versioning system will be confusing to use between insiders and stable. How does an insiders version is numbered without pre-release version tags?

- Since insiders version will be always latest, Markeplace shows insiders version of an extension like read me, version etc.,

## Approach 3: Insiders version tag (Marketplace hosted solution)

- Allow pre-release version tags `x.x.x-insiders` for extensions. 

- While browsing (VS Code or Marketplace) always show stable version by default unless requested to see insiders version.

- Provides a way (some UI) to the user to install the stable or opt in for insiders in VS Code. User can always switch between stable and insiders flavour of extension even after installing.

I would prefer Approach 3 which is correct IMO and has no drawbacks compared to other solutions. But this has a dependency on Marketplace to support this feature.

**Comment by davidanthoff** [comment]
Yes, I'm also much in favor of approach 3!

I think it would be nice if the pre-release tags would support things like `v1.4.3-alpha.2`, `v1.4.3-beta.3`, `v1.4.3-rc.1` etc. and then would automatically do the right thing. If that is too complicated, maybe just `v1.4.3-insiders.3` would do the trick as well. But it would be good if one could publish consecutive pre-release versions that are uniquely identified.

**Comment by DanTup** [comment]
I also prefer 3, since it seems weird if we have to keep revving non-preview version numbers for preview/insiders builds (however if it's likely to take significantly longer because it needs marketplace changes, I'd revise my preference ;-))

I'd also ask that `-insiders` builds can have additional build numbers (eg. `-insiders.1`, `-insiders.2`, `-insiders.3` etc.).

**Comment by eamodio** [comment]
I also strongly prefer 3

**Comment by sean-mcmanus** [comment]
As mentioned previously, our C/C++ extension implements this ourselves with our C_Cpp.updateChannel settings, so if the marketplace implements something similar it is important that our extension is able to:
1. Default to Insiders for VS Code Insiders.
2. Be able to show a prompt such as "Insiders version is available. Would you like to switch to the Insiders channel?: [Yes][Ask Me Later][Don't Show Again]" that can change the version.

Without those 2 features our Insider usage would be too low to be useful. Our code that does this is open source, but it's not packaged to be re-usable by other extensions.

As mentioned by the other users, some sort of 0.23.0-insiders, 0.23.0-insiders2 mechanism is desired: see https://github.com/microsoft/vscode-cpptools/releases .

The main disadvantage with our method is that our Insider versions are not available via the  "Install Another Version..." UI in VS Code...so if an API was available to make that fixable I'm not sure why we'd benefit from a marketplace change.

**Comment by TylerLeonhardt** [comment]
Without knowing the true architecture of the VSCode marketplace, I can only guess... But since vsix's are very similar to nuget packages, my guess is that the marketplace (VSCode, VisualStudio, etc) is some fork of NuGet.org (or NuGet.org is some fork of the marketplace). Also, the PowerShell Gallery is a fork of NuGet.org. 

Both NuGet.org and PowerShell took approach 2. My guess is because of the risk of doing approach 3 and making sure that we are still backcompat. Imagine trying to replace a bunch of SemVer v1 code with SemVer v2... There could be a ton of side effects.

With all that said, I would prefer approach 3 because it's the technically correct approach... But my understanding is that that's a different team and this would be a large ask so it's tough on that front.

Im voting for Approach 2 only because I think a lot of Approach 2 is needed for Approach 3 (like the toggle in vscode to show/hide insiders) and it will take the least amount of time and coordination.

**Comment by sandy081** [comment]
I think with Approach 2 extension authors will have troubles in version their stable and insiders versions. I really do not know what the model would be for versioning stable and insiders with `major.minor.patch`. For example if the current stable is `1.1.0` what is the next insiders version?

**Comment by DanTup** [comment]
If you went with 2, I'd probably just use even minor versions for stable releases and odds for insiders to keep things in order:

```
1.0.0 <-- stable
1.1.1 <-- insiders version of 1.2
1.1.2 <-- insiders version of 1.2
1.2.0 <-- stable
1.3.1 <-- insiders
1.4.0 <-- stable
```

It's not perfect since you can't have insiders versions of patch releases, but that doesn't seem like a *huge* issue IMO.

(I'm not saying I'd like 2, but if the difference between delivering that and 3 is a long time - it's been over 2 years since I opened this already - then I'd take it!)

**Comment by TylerLeonhardt** [comment]
I agree with @DanTup. That seems reasonable if it means we get the feature a lot sooner.

With the PowerShell extension, we recently followed the Python extension and switched to date-based versioning like `2019.8.0` where it's `year.month.buildThatMonth`. Leveraging that versioning scheme, with the `preview` property in the package.json determing prerelease (aka Approach 2) would work nicely for us.

P.S. We switched to that versioning scheme because we accidently shipped the preview extension (`2.0.0`) to the stable extension... It was a very painful experience. We couldn't risk using Semver anymore having 2 extensions to manage.

**Comment by davidanthoff** [comment]
If we went with 3), one could have a concept of a "channel': one could have a `nightly` channel, a `beta` channel (that is more stable and has fewer releases) etc.

The versioning scheme could be `v1.20.3-nightly.4` and `v1.20.3-beta.2` etc.

At the end of the day, I think that is what I'd _really_ like to have and it seems to be really useful in lots of other contexts (Windows, Edge etc.)

**Comment by janosh** [comment]
This would a great feature addition! Any timeframe on when this might land?

**Comment by vchuravy** [comment]
I seem to be getting different versions depending on whether I am on VSCode Insiders

```
Name: C/C++
Id: ms-vscode.cpptools
Description: C/C++ IntelliSense, debugging, and code browsing.
Version: 0.26.2-insiders2
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools
```

```
Name: C/C++
Id: ms-vscode.cpptools
Description: C/C++ IntelliSense, debugging, and code browsing.
Version: 0.26.1
Publisher: Microsoft
VS Marketplace Link: https://marketplace.visualstudio.com/items?itemName=ms-vscode.cpptools
```

But it looks like `cpptools` manually implemented that feature...
https://github.com/microsoft/vscode-cpptools/pull/3221

**Comment by sean-mcmanus** [comment]
@vchuravy Yes, and that's the default behavior, but it can be overridden via our C_Cpp.updateChannel setting. There's a bug preventing it from working for remote machines though (https://github.com/microsoft/vscode-cpptools/issues/3833).

**Comment by vchuravy** [comment]
It would be fantastic if extensions wouldn't have to implement this feature, but VSCode could offer it by default.

**Comment by sean-mcmanus** [comment]
@vchuravy Yeah, I agree. Our implementation is open source if any other extension wants to do something similar, but we haven't packaged it into an easily re-usable npm module. Other than the remote bug, our insider system has worked really good for our extension, allowing us to find/fix important bugs before they get released to a wider audience (and allowing users to get updates earlier than our slower 1.5 month non-Insiders release cadence).

**Comment by davidanthoff** [comment]
Is there any hope that this will make it to the top of the list of the core dev team? I think given that this needs integration with the marketplace, it really can only be done by the core team.

Every time we release a new version of our extension, I wish we had this. In my mind this is single feature that has the most amazing potential to increase the quality of the entire package ecosystem.

**Comment by GustavoASC** [comment]
I would love this

**Comment by davidanthoff** [comment]
We now caved and are publishing a second version of the Julia extension that is an insider build. The whole situation is really not great because the normal and the insider extension now conflict with each other, but it is better than nothing.

If anyone is interested in the Azure pipeline we use that creates the insider build, take a look [here](https://github.com/julia-vscode/julia-vscode/blob/a7248f87cc9a82482b1975b82f4f8656123a78f7/azure-pipelines.yml#L201). It makes the necessary modifications to the `package.json` and swaps out the README.

I still _really_ hope that at some point we'll get a proper solution for this...

**Comment by DanTup** [comment]
From reddit: https://www.reddit.com/r/vscode/comments/fakre5/hey_extension_authors/fj0r2az/

> I can speak to beta extensions. This is one of the most requested features for the marketplace and we're actively working on this one.

**Comment by moshfeu** [comment]
Having 2 extensions - the regular and the insider one raise a problem so the user can't have both of them. I've not digging into it but probably the issues are with duplicate ids of commands, views etc.
In principal, we can run a task to replace the ids with insider ids but it's very uncomfortable.

**Comment by J-Fields** [comment]
@moshfeu I'm eagerly awaiting beta releases as a first-class citizen, but as a workaround for your issue, you should be able to have the regular version look for the insider version (with [this api](https://code.visualstudio.com/api/references/vscode-api#extensions)) and disable itself if the latter is installed.

**Comment by moshfeu** [comment]
Thanks @J-Fields . The conflicts coming from the "contributors" - for example, buttons and commands. I'm not sure how using the API can solves this 🤔

**Comment by misak113** [comment]
This would prevent a lot of headaches to extension users :)
I'd prefer the possibility for the user to easily rollback to the latest stable version if he accidentally opted-in ;)

**Comment by anburocky3** [comment]
It would be cool like `Google Play Store` So the entire journey from alpha to production would be covered here.

**Comment by isidorn** [comment]
@sandy081 and me have been thinking about this topic and we understand that there are a lot of problems with the current approach of publishing two extensions: confusing for users, inconsistent across extensions, painful to manage both. We believe that the current separation of insiders and regular into two extensions does not make sense. Here's our proposed solution:

1. `vsce` will add support for an `--insiders` flag in the `publish` step. Such that extension authors could mark their `vsix` as `insiders`
2. When users uses VS Code Stable to install an extension, the user would get the latest non insider version of the extension. For VS Code Insiders the user would get the latest insiders version of the extension.
3. Users could manually switch from a Stable extension version to an Insiders and vice versa. Extensions will still auto-update respecting their "quality" - insiders extensions would update to latest insiders, stable only to stable extension versions
4. Insiders extensions will be decorated specially in the extensions view so users are not surprised by this
5. It is up to extension authors to deprecate their old separate "Insiders" version of the extension and all settings which might control this (C++ and Python). Either by using notifications or something else. We are willing to help with this transition.
6. Long term (upstream dependency): MarketPlace will add support for pre-release versions (e.g. 1.60.2-insiders). Once that is done we might deprecate the `--insiders`. And then the rendering on the Marketplace site will show the latest stable extension version. Currently it will show the latest "insiders" version. Which is fine for now - since most of the installation is done from VS Code app.

Please let us know what you think about this proposal. We are looking forward to the feedback. Thanks 👏 

Authors of extensions that have nightly versions @eamodio @connor4312 @alexr00 @aeschli @brettcannon @davidanthoff @dantup @wardengnaw @polinasok @hediet @greazer @TylerLeonhardt

**Comment by hediet** [comment]
@isidern I like the proposed solution!

Would vscode-stable users who installed an insider-extension also be updated to stable-builds of that extension if the stable build is newer than the insider build? (I think this makes a lot of sense)

In this case, those users should still get the newest insider build of the extension once it gets published. Thus I think the decision what to update to should not be based on the stability of the installed extension, but rather be a separate (user-configurable) flag.

**Comment by sandy081** [comment]
@hediet 

Thats a good point to note and Yes, we will be doing that - So if users opt to insiders/stable version of an extension then we will always have the best latest.

**Comment by chrmarti** [comment]
@rchiodo Mentioned using the `"engines"` property to limit a release to VS Code Insiders, not sure everyone is aware of that option:
```json
    "engines": {
        "vscode": "1.59.0-insider"
    },
```

**Comment by hediet** [comment]
> @rchiodo Mentioned using the `"engines"` property to limit a release to VS Code Insiders, not sure everyone is aware of that option:
> 
> ```json
>     "engines": {
>         "vscode": "1.59.0-insider"
>     },
> ```

But `1.60.0` is newer than `1.59.0-insider`, right? At least by the semver spec.

**Comment by isidorn** [comment]
Updated the proposal based on feedback (added drawbacks) and simplified that Stable Extensions would get downloaded for any VS Code flavour and users would have to opt into getting insiders extensions. We can later change this that by default VS Code insiders gets Insiders extensions. This is out of scope for this proposal.

**Comment by brettcannon** [comment]
What is the plan for user discoverability when it comes to this? With a separate nightly extension you get the SEO for your insiders version by showing up as a separate search result. Bundling under the same extension ID means you lose that. So would users then be expected to check every extension to see if there's an insiders version they can switch to? Or would extensions come up with their own way to "advertise" their insiders version to users and ask them to click the right thing to switch? Or would there be an `execCommand` we could use as part of a notification prompt to help users flip over to insiders?

**Comment by jakebailey** [comment]
The proposed plan would work well for Pylance; I designed our versioning to follow semver (technically, calver) with this eventuality in mind. Our stable releases are like `2021.9.1`, and our insiders are like `2021.9.2-pre.1` (insiders builds are always semantically newer than stable builds).

I was worried about the case where insiders users update to stable builds, when we want them to move up to other insiders builds, but I this already happens now as VS Code can update the extension to stable before we get a chance to manually install the VSIX. When a stable build of Pylance is release, we release a corresponding insiders build, so it's probably rare that anyone checks for an update during that small period and would work better than our manual daily checks.

**Comment by rchiodo** [comment]
I was hoping the default for VS code insiders would be to install the latest insiders (or stable if newer build)

Then it would happen automatically. The special ask the user if they want to opt in means nobody will end up using our insiders bits.

Right now we're using the engines property to accomplish just that. All VS code insiders users get our insiders version automatically (or stable version if it's newer).

**Comment by sean-mcmanus** [comment]
@isidorn VS Code Insiders defaulting to the latest Insiders extension is also a "requirement" for our extension to switch to this mechanism (instead of our current custom one) -- otherwise, Insider usage would be "too low" (we could probably provide telemetry data on this if you want).

**Comment by jakebailey** [comment]
Similar for Pylance; our custom insiders system also respects the Python extension's insiders setting, which gets us a more reasonable number of users. It's hard to make this discoverable.

**Comment by connor4312** [comment]
I think we should not talked about "newer" or "latest" builds, but _higher versioned_ builds.

>  Irrespective of being on insiders/stable version of an extension users we will always have the best latest ie., if stable is latest than insiders then always stable.

This is not a good behavior in my opinion. For example, often in recovery releases I'll go back and publish, for example, a `1.60.2` release even when the `1.61.0-insiders.1234` is available. In this case users on 1.60.1 should be updated, but those on the nightly should not be.

I think we should just endeavor to follow semver here; behavior in semver is already known and expected by, at minimum, all developers who work with JavaScript.

**Comment by sean-mcmanus** [comment]
@connor4312 I believe all references to "newer" and "latest" do mean "higher versioned" builds, and the release of 1.60.2 would not affect the 1.61.0-insiders if it's released later, i.e. it's just a word choice difference.

**Comment by connor4312** [comment]
Yea, just want to clarify that there should not be any chronological implication when we talk about version numbers. I thought there was currently a bug in the marketplace where that was assumed, but it looks like it's actually a separate issue (extension auto update does always only check for the highest versioned package, but if its engines.vscode is not compatible with the current vs code version the auto update stops working; I remembered and thought the problem was that it only returned the most recently published version)

**Comment by heartacker** [comment]
if insider avaible, IMO, I want this:
before installed:
![image](https://user-images.githubusercontent.com/1876302/132930641-d273ea2b-a6d3-4da0-9f76-7c79a6476735.png)

after installed:
![image](https://user-images.githubusercontent.com/1876302/132930660-d9aa8370-6438-4069-a3ed-c0116568a6e4.png)

**Comment by DonJayamanne** [comment]
> I think we should not talk about "newer" or "latest" builds, but higher versioned builds.

I'd like to add, the current behavior works very well for the Jupyter extension. We have releases for both stable and insider builds, and we ship recovery builds for stable every now and then, and insider builds on a daily basis

For Jupyter extension, if used wants insider builds, we just ask then to use the vscode insider builds (ie all vscode insider users automatically get three insider build of Jupyter extension). We haven't had any request from users to have insider builds of the extension in the stable version of vscode.

**Comment by DanTup** [comment]
> If a version has been published as insiders, it can not be promoted to stable, a new version has to be pushed. Can create strange version history for extensions

To clarify - does this mean _version numbers_? Eg. If I publish a 1.25.0 with `--insiders`, I cannot then also publish 1.25.0 without `--insiders`? Or does it just mean that we have to push the file again and can't just flip a flag? I'm assuming it's likely the former if the Marketplace isn't handling this and VS Code is instead tracking which versions were marked as insiders? (out of curiosity - how is this working is vsce just adding some additional metadata to the vsix?).

I think having the marketplace show the Insiders version number might be confusing, but it sounds like that's unavoidable without the Marketplace being updated. I'd certainly live with it to have Insiders versions :-)

I agree with other users though - having Insiders users automatically get Insiders by default would be nice to improve usage (or at least make the install view have the Insiders option more prominent).

**Comment by sandy081** [comment]
@brettcannon Your discoverability concerns are valid and we will definitely address this while coming up with UX mock ups. Our current thinking is to let VS Code handle this and not by extensions, so that users will get unified consistent experience.

@connor4312 Just to be clear about current proposal - **We are not changing current version semantics ie., extension still supports only `major.minor.patch`. We defer supporting prerelease version tag until Marketplace supports it.** Your scenario of publishing 1.60.1 when there is 1.61.0-insiders is supported only then. With the current proposal you still have to follow existing `major.minor.patch` version semantics but you can tell VSCE / VSCode that a specific version is a an insiders/prerelease one. Unfortunately you have to release a newer/higher version for your insiders, if you have a stable release. So in your case, you have to publish 1.60.1 as stable and 1.60.2 as insiders. This is mentioned as a drawback to the proposal and I agree that it does not sound good and adds some strange version gaps but this temporary and will be lifted once MP supports prerelease tags.

> To clarify - does this mean version numbers? Eg. If I publish a 1.25.0 with --insiders, I cannot then also publish 1.25.0 without --insiders? Or does it just mean that we have to push the file again and can't just flip a flag? I'm assuming it's likely the former if the Marketplace isn't handling this and VS Code is instead tracking which versions were marked as insiders? (out of curiosity - how is this working is vsce just adding some additional metadata to the vsix?).

@DanTup  your assumption is correct. Yes, with current approach VSCE/VSCode are just tracking the versions marked as insiders. Once MP supports prerelease versions we will move to it. Yes, we will be adding this as metadata to vsix for now.

**Comment by brettcannon** [comment]
What would the MP link to for downloading? The Python extension gets sideloaded a lot due to aggressive firewalls blocking VS Code hitting the MP for updates, so having the download link in the MP be the stable release would be good. Making the link just the highest versioned thing, insiders or not, can cause issues for extensions like Jupyter where users don't realize that the thing they are downloading from the MP is actually the insiders build and not the stable build.

**Comment by DanTup** [comment]
@brettcannon my understanding is that there will be _no_ changes to the marketplace in the short-term for the proposal above, so it _will_ just show the highest version number regardless of stable/insiders (because it won't know which are stable and which are insiders - only vsce and VS Code will understand that).

It's definitely not perfect, but if it's really likely to be 18 months for the Marketplace to get support, I'd rather have this. It's been almost 5 years since I opened this issue 🙃

**Comment by eamodio** [comment]
Before I created a separate Insiders extension for GitLens (and the same for GitHub Repositories) I think I would have (for the most part) loved this proposal, but now I feel there are significant issues with using the same extension id for both a stable and unstable build, at least in how VS Code handles data for them.

Having shared data storage for stable and unstable extensions, can be hugely problematic, not only for backward compatibility, but now you will have to deal with users switching back and forth between versions -- and with synced data this can get even more problematic. Granted extensions internally could try to deal with this, by storing their own data in separate stable vs unstable buckets, but that adds even more complexity (and must be understood before you hit the problem).

VS Code could, of course, store stable vs unstable data in separate buckets automatically and that might address most of my concerns.

Outside of the data concerns, I also feel like versioning will be challenging -- not only just getting the right thing published, but also for error reporting (since the main parts of the version are going to be either the same of very close). I definitely like how today both GitLens & GitHub Repositories use standard versioning `1.5.0` while the insiders version are `YYYY.MM.DDHHMM`. Which make it easy to tell the difference, but also how recent your insiders build it (since they are basically nightly builds). Maybe the `--insiders` or maybe better yet a `--prerelease-channel <name>` flag could not only specify the channel, but also affect the versioning structure -- so maybe all prereleases are versioned by the date.

Also if we think of this concept all as pre-release "channels" where the user can choose the channel they want to be in, it will be something easier to manage for authors, users, and the marketplace. On the marketplace, you could then show version histories by channel, download/upgrade by channel, etc. And maybe the number of channels and names are publisher controlled -- or maybe they are something we try to standardize on. And maybe vscode insiders get, by default, opt-ed into a pre-release channel, which they can opt-out of individually or globally. But in this way we aren't just thinking about version numbers -- version numbers are a secondary consideration after the channel.

Anyway that my 2c.

**Comment by brettcannon** [comment]
@DanTup for us it's more of a question as to whether we use this mechanism or publish a separate nightly extension for the next 18 months. We have a lot of users who sideload the Python extension, so having that "download" link in the MP point to something that isn't our stable release will be a support burden as people will inevitably open issues about getting our insiders build from there no matter how well we try to document where to actually download from.

**Comment by DanTup** [comment]
@brettcannon Yep, I understand. Presumably the proposal above doesn't prevent you from continuing as you are today, and consolidating things when the marketplace is updated.

For me, having a second extension would actually mean having a 3rd and 4th (Dart + Flutter - the Flutter extension has a dependency on Dart, so there couldn't be a Dart Insiders without also a Flutter Insiders) and I'd have to deal with users potentially having both sets installed. Having a single-extension solution would definitely be better for me even with the compromises above.

**Comment by isidorn** [comment]
Thanks a lot for feedback 👏  This week we plan to listen to your suggestions and to discuss this with the team. Next week we should have a definite decision if we will go down this road. Though let me try to answer some questions:

1. Good point by @eamodio to use different storage if extension is stable or insiders. This makes sense and if we implement the proposal @sandy081 and me will explore this
2. I have contacted the Marketplace team to see how much effort it would be for the Marketplace site to show the latest "non insider" version instead. I'll update here once I hear back from them.

**Comment by brettcannon** [comment]
@DanTup we just got feedback similar to your dependency explosion issue from @jakebailey as Pylance depends on our Python extension.

I expect we will go with whatever @isidorn and @sandy081 come up with and just be as clear as possible in our docs about where to download a stable VSIX to side-load (if the Marketplace team doesn't come back with "good" news for @isidorn 😉).

**Comment by sandy081** [comment]
All, please note the following change in the [proposal](https://github.com/microsoft/vscode/issues/15756#issuecomment-916728348) 

Given that MP supports publishing smaller/older versions (given that the version is not published already), Extension authors can still release insiders and stable (including patches) in parallel. One recommendation is to use patch for insiders versions but extension authors are free to use any numbering model given that they use only major.minor.patch semantics. Example:
- Assume that you have a stable released version 1.0.0 and want to release next insiders version.
- You can release next minor insiders versions as follows: 1.1.0, 1.1.1, ....... , 1.1.25 (marking them using `--insiders flag through vsce)
- You can promote above insiders version to stable version by releasing 1.1.26 and start working on next insiders versions - 1.2.0... just like above.
- If you want to fix a bug in previous stable ie., 1.1.*, you can fix the bug and release 1.1.27.

This will allow you to continue releasing insiders and at the same time you can release stable patches without affecting insiders version.

**Comment by isidorn** [comment]
We really appreciate the feedback and we will make a final decision next week whether to go ahead with the proposal.
In the meantime two more questions that I would like to get your feedback on:
1. Should VS Code insiders users gets insiders extensions by default? Or the insiders users need to opt-in to the insiders extensions?
2. If a user is on an Insider extension, and a new Stable version of that extension is released that has a higher version than the insider version. Should the user be updated to the just released stable? In other words opting to insiders does the user always get highest version even if that version is stable or not?

**Comment by chrmarti** [comment]
1. This might depend on the target stability of an insiders extension. If that's more on the experimental side, opt-in makes sense, if it's on the release-candidate side, opt-out will work better because it ensures a larger number of users will get it.
2. Extensions might enable experimental features in insiders versions, therefore insiders should stay on insiders also when there is a more recent stable version. (This seems similar to VS Code Insiders.) We might need a way for extensions to say when the insiders channel is discontinued, otherwise users might end up installing an out-of-date version.

**Comment by DanTup** [comment]
1. I'd probably prefer them to get Insiders by default, but I think @chrmarti makes some good points. Even if they have to opt-in, that still seems better than what we have today (and the two-extension thing is ofc already opt-in).
2. I think Insiders should stay on Insiders to avoid confusion, but again I don't feel too strongly about this. It's likely after a stable release I'd be making an Insiders release with a higher version number anyway.

**Comment by gjsjohnmurray** [comment]
1. Maybe allow the extension to decide?

**Comment by rchiodo** [comment]
1. I'd prefer insiders by default. Opt-in means a lot less insiders users for us and more confusion when asking people to try out fixes. Right now we just say if you want to try this fix, install VS code insiders.
2. I think I'd prefer insiders stay on insiders for the reasons that @chrmarti explained. Insiders bits can sometimes have experimental stuff that we're only trying in insiders and even if it doesn't, all fixes that end up in stable are always in insiders first for us anyway (so even if stable is newer, insiders will quickly follow).

**Comment by sean-mcmanus** [comment]
1. Our extension would highly prefer VS Code Insiders to get Insiders extensions versions by default; otherwise, Insider usage would drop too much.
2. Our extension would prefer Insiders to use the highest version, whether insiders or non-insiders, since our non-Insiders with a higher version is always a newer version than the previous Insiders -- that behavior could be overridden by extensions by publishing an insiders with a higher version as discussed in https://github.com/microsoft/vscode/issues/15756#issuecomment-919784379 and "experimental" features that are not desired in the latest non-Insiders could potentially be enabled via a setting instead of relying on a vsix version. This would allow our extension to publish 1.7.0-insiders, then 1.7.1, and then 1.8.0-insiders later on; otherwise, we'd have to publish 1.7.1 and 1.8.0-insiders simultaneously with identical bits, which would be unnecessary, although not a major issue.

A setting to allow extensions to opt-in to the different update behaviors would be fine too.

**Comment by jakebailey** [comment]
1. For Pylance, it would be beneficial to have insiders automatically get insiders extensions, but this isn't something we have now, and I'll personally be happy with any sort of increase in this userbase, be it from it being auto-installed or the UI being extended to show that insiders versions are available (which is going to have to happen anyway). I can see this potentially causing an issue from the POV that people trying out VS Code insiders will not only get VS Code's unstableness, but any unstableness from every extension they have installed that also has insiders.
2. I don't know if it makes sense for extensions to be upgraded from insiders to stable; if storage is separated out as in https://github.com/microsoft/vscode/issues/15756#issuecomment-919359976, this will probably cause oddities as users move back and forth within the channels. For example, we might intend to show a popup once (then never again), except that moving between stable/insiders will effectively clear that status. In Pylance's case, every time we publish a stable version, we also automatically publish a new insiders build that is semantically newer (so 2021.9.2 is equivalent to 2021.9.3-pre.1), so I'm not too worried either way.

**Comment by sandy081** [comment]
> I don't know if it makes sense for extensions to be upgraded from insiders to stable; if storage is separated out as in Allow extensions to publish beta releases and users to opt-in to them #15756 (comment), this will probably cause oddities as users move back and forth within the channels.

@jakebailey I do not think we would be having separate storages for stable & insiders of the extension. As of now we would like to go with current model - one storage per extension. Having two storages will be an advanced feature request going forward.

**Comment by sandy081** [comment]
> Maybe allow the extension to decide?

@gjsjohnmurray Yeah, extensions can do that by using vscode engine property in package.json

**Comment by connor4312** [comment]
For (2) my vote is in the court that Insiders users be updated to Stable if the most recent stable version > insiders. This matches the semver model to avoid inventing new mechanics for VS Code's versioning. Perhaps the `ExtensionContext` has a `updateChannel` enum that lets the extension know whether this is a stable or insiders user in order to enable/disable features?

**Comment by CsCherrYY** [comment]
Overall, the proposal is very helpful for the Java extension since we need an insider version to expand the visibility of our beta features. For the two questions:

1. I suggest VS Code insiders users could get insiders extensions by default. Since Java extension doesn't have an insider version yet, this will help our insider extensions used by more users and they can also have a change to switch back to the stable versions if needed.
2. When `--insiders` becomes available, we plan to release a new stable version along with a new insider version to keep the newest semantic version of Java extension is always the insider version, just like what VSCode insiders and Pylance do (mentioned before). Thus, Java extension will not meet this problem, I'm OK with both choices.

**Comment by isidorn** [comment]
Thank you very much for your great feedback. After a milestone of house keeping and thinking about all of this @sandy081 @misolori and me plan to do the following this milestone:
* Continue to use the current `major.minor.patch` version semantics but provide a way to extension authors to mark a version as insiders by introducing a flag `--insiders` to `vsce` during a publish step
* If a user opts to insiders extension the user will always get the highest version even if that version is stable
* For VS Code Insiders users by default we will offer users to get the Insiders extensions (image attached). We will use a dropdown so users can easily choose to go with the stable version instead. On VS Code stable users can still choose but the default option will be to install the Stable version of the extension.
* If an extension has an Insiders version available we will always render the Green badge in the upper right corner to indicate this both in VS Code Stable and Insiders
* There will be an easy way for users to switch between Stable and Insiders versions of an extension. Both for VS Code Stable and VS Code Insiders
* We will provide migration support for extensions. For example, when the separate Git Lens Insiders extension gets deprecated VS Code will automatically install the Git Lens extensions and switch it to the Insiders version. We will also transition the extension state. Extension authors will have to contact us for this explicitly
* Marketplace will render the latest non Insiders extension version on the Marketplace site

![Screenshot 2021-11-12 at 11 39 57](https://user-images.githubusercontent.com/1926584/141453865-468fc886-742a-46bf-a4ad-d11bbc6eb2c7.png)

**Comment by DonJayamanne** [comment]
@isidorn thanks for the update

> We will provide migration support for extensions. For example, when the separate Git Lens Insiders extension gets deprecated VS Code will automatically install the Git Lens extensions and switch it to the Insiders version. We will also transition the extension state. Extension authors will have to contact us for this explicitly

The above didn't make much sense. Your are talking about migrating insider gitlense to insider.

**Comment by isidorn** [comment]
@DonJayamanne thanks. I have tried to clarify this by editing that sentence and by adding links to clarify which extension I am exactly talking about in what part of the sentence. Hope it is clearer now.

**Comment by CsCherrYY** [comment]
Thank you all for the great work. Here I have another question, can extension have a way to know whether it's an insider version itself in runtime? Thus, we can use different APIs easily.

**Comment by jks-liu** [comment]
> Thank you all for the great work. Here I have another question, can extension have a way to know whether it's an insider version itself in runtime? Thus, we can use different APIs easily.

I don't think it's a good idea. Same extension with same version should has same behaviour, no matter insider or not.

I think for your case you can get version string in runtime, and use different API according only to the version.

**Comment by DanTup** [comment]
@isidorn sounds neat! I presume if we push a stable version that's higher than the last insiders version, everyone (including insider-extension-version users) will be upgraded to it, but VS Code will remember those that were on the "insiders channel" so if we then push another Insiders version, those users would update back to that?

**Comment by sandy081** [comment]
Yes, that's correct.

**Comment by eamodio** [comment]
@isidorn Thanks for the update.

I think I'm good with most of this, but my biggest concern is when a user transition from an insider's version back to stable. Thinking mainly of data and settings migrations that could have happened that could now cause issues with the stable version (leading to a broken or unstable "stable" version). For stored data this can be mitigated more easily by storing the data for stable vs insiders differently (e.g. scoping the key), but with settings it gets a bit trickier. Maybe this is mostly a non-issue as dealing with these issues would only arise in a small set of extensions, but it might be good to have some documentation around best-practices to handle/avoid them.

My other thought is around versioning, but I think it might be a non-issue (though I haven't thought through all the scenarios). Most Insiders/Nightly extensions I've see use a date/time versioning model, of which I think is highly valuable to both the extension and the user, and I think give the rules outlined above an extension could still utilize this versioning. Since Insiders versions using `YY.MM.DDHHMM` (or similar) would always be greater than the stable version, so you'd never get "upgraded" from Insiders to Stable (which imo can be desirable). Are there any other issues with try to keep that versioning scheme (if an extension wanted) that you all can think of?

A couple of thoughts/clarifications on the UI mocks:
 - In the images of the 2 versions of GHPR -- they both seem to say they are the same version. I'm guessing (hoping) that was just a mockup issue, and if you were on version `1.5` on stable, and you switch to insiders (and the latest is `2021.11.151010`) then `2021.11.151010` will be shown as the version in the UI? 
 - It seems redundant to have the "This extension has an Insiders version available" message when you are on the insiders version -- almost makes it feel a bit broken.
 - I think there should be a badge or something applied to the extension in the list when you are running an insiders version -- easier to scan for -- I think it probably should warrant a badge on the extension overview page too
 - I think it would be nice for extensions to be able to provide an insiders icon image if they want to further differentiate the editions
 - I think the terminology of "Switch to [Insiders|Stable] Extension" on the button could be improved by calling them Channels (or Editions, though I much prefer channels). I think it could also help users better understand what they are opting into. Other platforms use the channels metaphor.

> For VS Code Insiders users by default we will offer users to get the Insiders extensions (image attached). We will use a dropdown so users can easily choose to go with the stable version instead. On VS Code stable users can still choose but the default option will be to install the Stable version of the extension.

Can you show a mock of what this would look like for a first time install? Both in the list and in the overview page?

**Comment by isidorn** [comment]
@eamodio thanks a lot for the feedback. Let me try to answer your questions:

Yes, we should think more about what transition from Stable to Insiders extensions means for data and settings and we should document it. @sandy081 probably has more thoughts here

No, I do not see other issues with that versioning scheme.

1. Yes, this is a mockup issue
2. I agree. We'll improve this in the real version
3. There is a green circle badge in the top right which shows this. We can fine tune this
4. This is a fair feature request, not sure if this might be possible to do already if an extension publishes a different icon every time it publishes the `insiders` editions. I understand this is a manual process and we might need something more straight forward. Good feature request after we ship this
5. I like the word "channel". We could use "Switch to Insiders Extension Channel" / "Leave Insiders Extension Channel". Might be too long. We decided to not use the word "stable". @misolori for thoughts

First time install for VS Code Insiders would look something like this (mock still in polish phase)

![Screenshot 2021-11-16 at 11 24 59](https://user-images.githubusercontent.com/1926584/141968152-b8fbb383-1dad-443b-af39-7b2fb3d6f712.png)

**Comment by bobbrow** [comment]
@isidorn I've been thinking about the versioning more as we consider what it would be like to transition the C++ extension to this system.  Previously, if our upcoming release was, say, 1.5.0, we would have 1.5.0-insiders, 1.5.0-insiders2, etc, and then finally 1.5.0 (stable).  We don't do daily insiders releases and only start releasing them when new features are available for evaluation.  The way VS Code determined which extension to install was based some kind of "greater than" check, so our developers in the insiders program would automatically be upgraded to "stable" when the evaluation period for a release had completed.

It occurs to me that with this new way if only x.y.z versioning is supported, we'd have to do versioning one of two ways:

**`x.y.0` is the first stable**
Ideally, when we release a new version, the patch number is `.0`.  In order to achieve this, it seems we'd have to release insiders for `x.y` as `x.(y-1).z`.  This might be confusing because it doesn't indicate that new features are becoming available since only the patch version is changing.

Example:
Current stable release is `1.4.0`.  We release the first insiders as `1.4.1`, second insiders as `1.4.2`.  When stabilized we release `1.5.0`.

**`x.y.z` is the first stable**
In order to ensure that our insiders are upgraded to stable at the end of the evaluation period, we would start releasing insiders as `x.y.0`, then `x.y.1`, `x.y.2` for subsequent insiders releases. The first stable release would have a version number of `x.y.z`.

Example:
Current stable release is `1.4.3`.  We release the first insiders as `1.5.0`, second insiders as `1.5.1`, etc.  When stabilized we release `1.5.3` (or whatever the current patch number is).


The nice part about using the `-` in our versions currently is that it allows us to denote that new features are coming (increment the minor version number) and also release the first stable with a patch number of 0.

Can we continue to use `-` in our versions when publishing with `--insiders`, or how would you recommend we do our versioning going forward?

**Comment by eamodio** [comment]
@isidorn Thanks!

I like the idea of Switch/Leave the insiders channel -- I think you can shorten it to `Switch to Insiders Channel` / `Leave Insiders Channel` or even `Join Insiders Channel` / `Leave Insiders Channel`. 

Alternatively, rather than buttons, this could be a "[Release] Channel" dropdown that has a `Stable` (or `Release`) and `Insiders` options.

But I think in any of these options there needs to be some explanatory text as to what an insiders version/channel is.

**Comment by sandy081** [comment]
- I do not think we would like to handle any data partition or migration for stable and insiders versions of the extensions. VSCode would treat this as a normal update to higher/lower version. We will make it clear about this in the doc and also recommend such extensions to use separate storage keys (append with insiders for eg) for differentiating. 

- I think using **Channel** might send different message about this feature. Briefly, with this feature, VSCode is allowing users to install and use insiders/pre-release versions of an extension. For such users, VSCode will always install the latest version of the extension and this can be a pre-release or release version. So the name `Join/Leave the Insiders Channel` might convey that user will be always in Insiders channel even if there is latest stable. It is up to the extension to always have a latest insiders version and force users to be on insiders version but it is not mandatory. Yes, I agree that we have to improve on conveying this message in our UI. 

- Using patch version for pre-release tags is a temporary solution because Marketplace does not support pre-release tags at present. We (VSCode & Marketplace) have plans to support pre-release tags (SemVer) going forward and once it is available you can start using pre-release tags and follow the same release cycle as you are using now, ie., `1.5.0-insiders` -> `1.5.0-insiders2` -> `1.5.0`. Until then we recommend to use patch part for pre-release/insiders versions as mentioned in this [comment](https://github.com/microsoft/vscode/issues/15756#issuecomment-916728348)

CC @eamodio @bobbrow @isidorn

**Comment by bobbrow** [comment]
Thanks @sandy081

**Comment by Lemmingh** [comment]
> Briefly, with this feature, VSCode is allowing users to install and use insiders/pre-release versions of an extension.
>
> <https://github.com/microsoft/vscode/issues/15756#issuecomment-970482005>
> <https://github.com/microsoft/vscode/issues/15756#issuecomment-971450154>

What about still having just one "**Install**" button and adding a "**Receive Insiders**" switch next to it?

```text
Install | ☑ Receive Insiders

This extension has an Insiders version.
```

```text
Disable | Uninstall | ☑ Receive Insiders

You are receiving an Insiders version. This extension is enabled globally.
```

```text
Disable | Uninstall | ☐ Receive Insiders

An Insiders version is available. This extension is enabled globally.
```

**Comment by isidorn** [comment]
@Lemmingh thanks for your feedback, however we have decided on the current design, we are currently in the "polish wording" phase.

**Comment by isidorn** [comment]
We have pushed this feature to Insiders 🚀  
We are in the process of testing it. If all looks good end of this week I will write docs for extension authors on how to best adopt to this and I will share them here. 
[GitHub Pull Request extension](https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github) adopted to this, so for those who are eager to try out the flow can use VS Code insiders and that extension to see how it looks.

**Comment by DominicVonk** [comment]
It would be great if we can publish a pre release with the proposed apis enabled.

**Comment by isidorn** [comment]
We have released a VS Code stable that supports this 🚀 
Here are the extension author docs that should help you adopt to pre-release extension support
https://code.visualstudio.com/api/working-with-extensions/publishing-extension#prerelease-extensions

If there are any questions do not hesitate to reach out. Looking forward to feedback. Thanks a lot 👏

**Comment by DanTup** [comment]
@isidorn awesome! Two questions:

- what happens if we publish a pre-release version now, for users that haven't upgraded VS Code yet? Would they see the pre-release version as an update, or would it be filtered out on the server?
- in a comment above you said "Marketplace will render the latest non Insiders extension version on the Marketplace site" but on the page at https://marketplace.visualstudio.com/items?itemName=GitHub.vscode-pull-request-github it shows `Version | 0.35.2021120909` which looks like an insiders version (I see 0.34.0 in VS Code as the stable version). Is this a bug, or has the marketplace not been updated yet?

**Comment by isidorn** [comment]
@dantup good questions:
* `vsce` will require you to specify an engine property in your pre-release version so that older VS Code versions will not be able to use it. So users that haven't upgraded will not be offered the pre-release version
* Correct, MP has not yet done this work. This item tracks the work here https://github.com/microsoft/vsmarketplace/issues/264 In the end we decided to have more UI decorations to make it clearer to the user that it is a pre-release version. We should have something on the MP side by the time most of the extensions adopt to this (end of January). Let's continue the discussion on that issue for this.

**Comment by DanTup** [comment]
> `vsce` will require you to specify an engine property

Ah, perfect. Thanks! :-)

**Comment by CsCherrYY** [comment]
@isidorn Thanks for the hard work. We're looking forward to a way to publish Type Hierarchy feature as a preview feature and pre-release extension is a great choice. However, the related APIs are in proposed state (see https://github.com/microsoft/language-server-protocol/pull/1231#issuecomment-929296135 and https://github.com/microsoft/vscode-languageserver-node/blob/main/client/src/common/proposed.typeHierarchy.ts), enabled by `enabledApiProposals` and we can't publish it even to the pre-release channel.

The question is that we want to find a way to take advantage of the pre-release channel to make more users use our preview features with proposed APIs, is that possible? Or is there any consideration about this?

**Comment by isidorn** [comment]
@CsCherrYY there have been some discussions around this but nothing definite yet. We want to make sure that the `pre-release` extension support is adopted smoothly across our community before we decide to build any additional features on top.

**Comment by Christoc90** [comment]
> Toutes nos excuses si cela s'est déjà produit ; Je n'arrive pas à trouver de problèmes liés.
> 
> Ce serait cool de pouvoir publier des extensions sous une forme bêta (similaire à Code Insiders) afin qu'elles puissent être exécutées "en production" avec de vrais utilisateurs qui se sont inscrits. Cela réduirait le disque lors de la publication d'une extension car vous pouvez la transmettre à un plus petit nombre d'utilisateurs pour le tester (je n'ai probablement pas besoin de vous le vendre, je suppose que vous avez des initiés pour des raisons très similaires).
> 
> Je sais que nous pouvons empaqueter et distribuer les extensions nous-mêmes pour le moment, mais ce n'est pas aussi simple (par exemple, la mise à jour n'est pas automatique).
> 
> Je pensais à l'origine que le drapeau "Aperçu" aurait pu fournir cette fonctionnalité, mais à moins que j'aie oublié quelque chose, il ne semble pas le faire.

**Comment by davidanthoff** [comment]
Is there some way to find out from code whether the currently running extension was installed as a pre-release version or not? We want to send telemetry to different destinations, depending on whether a user is running a pre-release version or not.

**Comment by TwitchBronBron** [comment]
> Is there some way to find out from code whether the currently running extension was installed as a pre-release version or not? We want to send telemetry to different destinations, depending on whether a user is running a pre-release version or not.

A version cannot be both pre-release and not pre-release, so if you manage the extension, then you are in control of the process for launching pre-release versions and therefore know which versions are pre-release. If you follow the suggestions outlined in the documentation:

> we recommend that extensions use major.EVEN_NUMBER.patch for release versions and major.ODD_NUMBER.patch for pre-release versions. For example: 0.2.* for release and 0.3.* for pre-release.

then you can easily check the current extension version and act accordingly based on whether is an even or odd minor release.

**Comment by isidorn** [comment]
@davidanthoff exactly as @TwitchBronBron explained: you can look at your extension version and act accordingly.

**Comment by rchiodo** [comment]
There's a lot to read through in this issue and I wasn't sure what the expected behavior is with regards to 'insiders'.

Should insiders users automatically be updated the prerelease of an extension?

This isn't currently happening for the [Jupyter Extension](https://marketplace.visualstudio.com/items?itemName=ms-toolsai.jupyter).

I had to explicitly pick 'Switch to prerelease version'

![image](https://user-images.githubusercontent.com/19672699/148584097-a4ebc807-eba5-424a-a126-89239eed6e93.png)

**Comment by sean-mcmanus** [comment]
Uh oh...if VS Code Insiders users have to manually opt-in, we're likely to see our Insiders usage drop by a lot. @bobbrow Are you sure we should switch to the new mechanism if that's the case? And/or can we get VS Code to change the behavior first?

**Comment by isidorn** [comment]
We support automatically migrating separate extensions to the pre-release version of the main extension - extension authors just have to inform us they want this.
However since C++ and Jupyter have a special case and you do not have separate extensions we currently do not support automatic migration done by VS Code. However you can do the migration on your own as discussed here https://github.com/microsoft/vscode-cpptools/issues/8507#issuecomment-994178165

**Comment by gjsjohnmurray** [comment]
For me just now, when I used Insiders and searched for the GHPRI extension (which I know published pre-releases), the Install button defaulted to Pre-Release:

![image](https://user-images.githubusercontent.com/6726799/148587401-104ddc2d-ce21-4a32-9717-bbce39b56a94.png)

Doing the same with Stable defaults to a regular install:

![image](https://user-images.githubusercontent.com/6726799/148587621-cba0edc6-7bf0-4107-8dee-9fdf1703f2ea.png)

**Comment by rchiodo** [comment]
> We support automatically migrating separate extensions to the pre-release version of the main extension - extension authors just have to inform us they want this. However since C++ and Jupyter have a special case and you do not have separate extensions we currently do not support automatic migration done by VS Code. However you can do the migration on your own as discussed here [microsoft/vscode-cpptools#8507 (comment)](https://github.com/microsoft/vscode-cpptools/issues/8507#issuecomment-994178165)

@isidorn I'm not sure that's going to work? Won't it force all users to the latest? For me I didn't want the latest, but when I do, it should show me the latest is the prerelease version. That's the behavior we have right now. Users pick when they upgrade, but the recommended upgrade is the most recent insiders only build.

**Comment by isidorn** [comment]
@gjsjohnmurray correct.
@rchiodo yes, the insiders users will by default get the option to install Pre-Release. My previous answer was my misunderstanding of your question. For me this just works

![Screenshot 2022-01-07 at 19 12 18](https://user-images.githubusercontent.com/1926584/148588034-e44b3b7c-8ea8-4d16-af6b-d88d0abe5918.png)

**Comment by isidorn** [comment]
@rchiodo you would just do that code for your insider version of the extension. So that way you would transition only your insider users to the new pre-release version.

**Comment by rchiodo** [comment]
@isidorn  that's the behavior with the jupyter extension not already installed. But if it was already installed, the update notification won't pick prerelease.

And I believe the command suggested would force an update? Maybe I'm wrong about that.

**Comment by rchiodo** [comment]
I guess it won't be terrible to force prerelease if the extension isn't already in insiders for a specific version (one time thing).

**Comment by isidorn** [comment]
@rchiodo there is just one minor thing missing to have this flow super nice https://github.com/microsoft/vscode/issues/139528
We plan to tackle this in the next two weeks.

And yes. So my full suggestion is the following:
1. For your current Insiders jupyter version do the following
2. Show a persistent notification to the user, or a modal dialog "bla bla we moved to pre-release - click here to be migrated"
3. Once user clicks -> execute the code. Once #139528 is done all should be good

Hope I was more clear now. Thank you

**Comment by eamodio** [comment]
Also another way to ensure users don't automatically get switched to a non-insiders release (if that is what you want) -- you can have your insiders versions always be greater than stable version. For example in GitLens I am planning to keep my `YYYY.MM.DDHH` format.

**Comment by alexr00** [comment]
> Also another way to ensure users don't automatically get switched to a non-insiders release (if that is what you want) -- you can have your insiders versions always be greater than stable version.

GitHub Pull Requests and issues does the same thing. Before releasing a new stable, we will always release a new insiders with the increased version.
